metrics/stackdriver_exporter.go

```diff
@@ -53,8 +55,26 @@ var (
 	initClientOnce sync.Once
 	// kubeclientInitErr capture an error during initClientOnce
 	kubeclientInitErr error
+
+	// setStackdriverSecretLocation protects setting secretNamespace and secretName
+	setStackdriverSecretLocation sync.Once
+	// secretName is the name of the k8s Secret to pass to Stackdriver client to authenticate with Stackdriver.
+	secretName = DefaultSecretName
+	// secretNamespace is the namespace to search for a k8s Secret to pass to Stackdriver client to authenticate with Stackdriver.
+	secretNamespace = DefaultSecretNamespace
 )
 
+// SetStackdriverSecretLocation allows setting non-default values for the name and namespace of
+// the Secret that can be used to authenticate with Stackdriver.
+// The Secret is only used if metricsConfig.stackdriverClientConfig.UseSecret is true.
+// This can only be called once.
+func SetStackdriverSecretLocation(name string, namespace string) {
```

Comment:
Why is it important that this only happens once?

What happens if this gets called after the stackdriver exporter is created? It appears as though unless the metrics config itself changes, the exporter won't get re-created.

Reply:
It's not particularly important, but access to the state does need to be synchronized and thread-safe. So it was between a lock or a sync.Once. 

For security reasons, the values here shouldn't be user configurable and even controllers shouldn't really be modifying them throughout their lifetime, so sync.Once seemed appropriate.

Reply:
I'm concerned with @garron's comment about not updating until the metrics config is re-read.

Do we have a way to regenerate the stackdriver client after this Once is done?

Reply:
The intention here are that these values are essentially hardcoded, with the exception that different Knative consumers may want to set different values (i.e. "knative-serving" vs "knative-eventing"). If non-default options are not sufficient, consumers should set this function before beginning to export metrics, though I can add a comment to clarify that.

The set once is to emphasize that these should effectively be const values, we can't have external agents re-configuring the Secret name/namespace because of the queue-proxy security issue mentioned in the PR description.

Reply:
I switched it to a Lock to give more flexibility. pkg consumers can call UpdateExporter() after setting the secret location if they wish.

Users cannot configure the Secret location.